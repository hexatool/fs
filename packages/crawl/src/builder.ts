import CrawlerBuilder from './builder/index';

export default CrawlerBuilder;
