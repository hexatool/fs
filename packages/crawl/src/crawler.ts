import type { Dirent } from 'node:fs';
import { basename, dirname, resolve as pathResolve, sep } from 'node:path';

import invokeCallback, { type InvokeCallbackFunction } from './functions/invoke-callback';
import joinDirectoryPath from './functions/join-directory-path';
import joinPath, { type JoinPathFunction } from './functions/join-path';
import pushDirectory, { type PushDirectoryFunction } from './functions/push-directory';
import pushFile, { type PushFileFunction } from './functions/push-file';
import resolveSymlink, { type ResolveSymlinkFunction } from './functions/resolve-symlink';
import walkDirectory, { type WalkDirectoryFunction } from './functions/walk-directory';
import type { CrawlDirection, CrawlerOptions, ExcludePredicate } from './options';
import { checkOptions } from './options';
import { CrawlerQueue } from './queue';
import { cleanPath } from './utils';

export interface CrawlerState {
	options: CrawlerOptions;
	paths: string[];
	queue: CrawlerQueue;
}

export type ResultCallback = (error: Error | undefined, output: string[]) => void;

export class Crawler {
	private readonly callbackInvoker: InvokeCallbackFunction;
	private readonly direction: CrawlDirection;
	private readonly isSynchronous: boolean;
	private readonly joinPath: JoinPathFunction;
	private readonly pushDirectory: PushDirectoryFunction;
	private readonly pushFile: PushFileFunction;
	private readonly resolveSymlink?: ResolveSymlinkFunction | undefined;
	private readonly state: CrawlerState;
	private readonly stopAt?: string | undefined;
	private readonly walkDirectory: WalkDirectoryFunction;

	constructor(root: string, options: CrawlerOptions, callback?: ResultCallback) {
		checkOptions(options);
		this.isSynchronous = !callback;

		this.state = {
			// Perf: we explicitly tell the compiler to optimize for String arrays
			paths: [''].slice(0, 0),
			options,
			queue: new CrawlerQueue((error, state) => this.callbackInvoker(state, error, callback)),
		};
		this.direction = options.direction;
		this.stopAt =
			'stopAt' in options && options.stopAt ? this.normalizePath(options.stopAt) : undefined;
		this.callbackInvoker = invokeCallback(this.isSynchronous);
		this.joinPath = joinPath(root, options);
		this.pushDirectory = pushDirectory(options);
		this.pushFile = pushFile(options);
		this.resolveSymlink = resolveSymlink(this.isSynchronous, options);
		this.walkDirectory = walkDirectory(this.isSynchronous, options);
	}

	start(root: string, depth: number): string[] | undefined {
		const normalizePath = this.normalizePath(root);
		this.walkDirectory(this.state, normalizePath, depth, this.crawl);

		return this.isSynchronous ? this.callbackInvoker(this.state, undefined) : undefined;
	}

	private readonly crawl = (entries: Dirent[], directoryPath: string, depth: number) => {
		const {
			paths,
			options: { filters, exclude },
		} = this.state;

		const resolveSymlinks =
			'resolveSymlinks' in this.state.options && this.state.options.resolveSymlinks;

		this.pushDirectory(directoryPath, paths, filters);

		const files = this.state.paths;
		for (const entry of entries) {
			if (entry.isFile() || (entry.isSymbolicLink() && !resolveSymlinks)) {
				const filename = this.joinPath(entry.name, directoryPath);
				this.pushFile(filename, files, filters);
			} else if (entry.isDirectory()) {
				const nextDirectory = joinDirectoryPath(entry.name, directoryPath);

				if (exclude && exclude(entry.name, nextDirectory)) {
					continue;
				}

				if (this.direction === 'up') {
					this.pushDirectory(nextDirectory, paths, filters);
					continue;
				}

				this.walkDirectory(this.state, nextDirectory, depth - 1, this.crawl);
			} else if (entry.isSymbolicLink() && resolveSymlinks) {
				const path = joinDirectoryPath(entry.name, directoryPath);
				this.resolveSymlink!(path, this.state, (stat, resolvedPath) => {
					if (stat.isDirectory()) {
						const normalizePath = this.normalizePath(resolvedPath);
						if (exclude && exclude(entry.name, normalizePath)) {
							return;
						}

						if (this.direction === 'up') {
							this.pushDirectory(normalizePath, paths, filters);
						} else {
							this.walkDirectory(this.state, normalizePath, depth - 1, this.crawl);
						}
					} else {
						this.pushFile(resolvedPath, files, filters);
					}
				});
			}
		}

		if (this.direction === 'up' && directoryPath !== this.stopAt) {
			const nextDirectory = this.previousDirectory(directoryPath, exclude);
			if (nextDirectory) {
				this.walkDirectory(this.state, nextDirectory, depth - 1, this.crawl);
			}
		}
	};

	private normalizePath(path: string) {
		let copyPath = path;
		if (this.state.options.resolvePaths) {
			copyPath = pathResolve(copyPath);
		}
		if (this.state.options.normalizePath) {
			copyPath = cleanPath(copyPath);
		}
		const needsSeparator = copyPath[copyPath.length - 1] !== sep;

		return needsSeparator ? copyPath + sep : copyPath;
	}

	private previousDirectory(
		directoryPath: string,
		exclude?: ExcludePredicate
	): string | undefined {
		const dirName = basename(directoryPath);
		const nextDirectory = this.normalizePath(dirname(directoryPath));
		if (exclude && exclude(dirName, directoryPath)) {
			const stepNextDirectory = this.normalizePath(dirname(directoryPath));

			return this.previousDirectory(stepNextDirectory, exclude);
		}

		return nextDirectory;
	}
}
