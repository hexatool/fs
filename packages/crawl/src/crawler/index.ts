import { DirentAsyncCrawler, StringAsyncCrawler } from './AsyncCrawler';
import { DirentSyncCrawler, StringSyncCrawler } from './SyncCrawler';

export { DirentAsyncCrawler, DirentSyncCrawler, StringAsyncCrawler, StringSyncCrawler };
