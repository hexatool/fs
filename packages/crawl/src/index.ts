import CrawlerBuilder from './builder';

export default CrawlerBuilder;
