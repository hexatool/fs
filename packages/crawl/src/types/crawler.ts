import type { PathLike } from 'fs';

export default interface Crawler<TypeReturn> {
	start(path?: PathLike): TypeReturn;
}
