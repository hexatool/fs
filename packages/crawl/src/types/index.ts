import type Crawler from './crawler';
import type { CommonCrawlerOptions, CrawlDirection, CrawlerOptions, ReturnType } from './options';
import { directions, resultTypes } from './options';

export const apiTypes = ['async', 'sync', 'stream', 'iterator'] as const;
export type ApiType = (typeof apiTypes)[number];

export const emitEvents = ['close', 'data', 'end', 'error', 'pause', 'readable', 'resume'] as const;
export type EmitEvents = (typeof emitEvents)[number];

export type CallBack<Result> = (error: NodeJS.ErrnoException | null, result: Result) => void;

export type { CrawlDirection, ReturnType };

export { CommonCrawlerOptions, Crawler, CrawlerOptions, directions, resultTypes };
