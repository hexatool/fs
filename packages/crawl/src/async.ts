import type { Dirent } from 'node:fs';

import builder from './builder';
import type { CrawlerOptions, DirentCrawlerOptions, StringCrawlerOptions } from './types';

export type { CrawlerOptions };
export default function crawl(options: DirentCrawlerOptions): Promise<Dirent[]>;
export default function crawl(options: StringCrawlerOptions): Promise<string[]>;
export default function crawl(path: string, options: DirentCrawlerOptions): Promise<Dirent[]>;
export default function crawl(path: string, options: StringCrawlerOptions): Promise<string[]>;
export default async function crawl(
	pathOrOptions: CrawlerOptions | string,
	options?: CrawlerOptions
): Promise<(Dirent | string)[]> {
	const path =
		typeof pathOrOptions === 'string'
			? pathOrOptions
			: 'direction' in pathOrOptions
			? undefined
			: pathOrOptions;
	const opts =
		typeof pathOrOptions === 'string'
			? options
			: 'direction' in pathOrOptions
			? pathOrOptions
			: options;
	if (!opts) {
		throw new Error(`options could not be undefined`);
	}

	const b = builder[opts.direction]();
	if (opts.returnType === 'Dirent') {
		return b.withDirent().async(path);
	}

	return b.async(path);
}
