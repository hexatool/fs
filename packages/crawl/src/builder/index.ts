import type { Dirent } from 'node:fs';
import type { Readable } from 'node:stream';

import {
	DirentAsyncCrawler,
	DirentSyncCrawler,
	StringAsyncCrawler,
	StringSyncCrawler,
} from '../crawler';
import { DirentIteratorCrawler, StringIteratorCrawler } from '../crawler/IteratorCrawler';
import { DirentStreamCrawler, StringStreamCrawler } from '../crawler/StreamCrawler';
import type { Crawler, CrawlerOptions } from '../types';
import type Builder from '../types/builder';
import type { CrawlerDownOptions } from '../types/options';

abstract class CommonCrawlerBuilder<T extends CrawlerOptions, O extends Dirent | string>
	implements Builder<O>
{
	protected readonly options: T;

	protected constructor(options: T) {
		this.options = options;
	}

	abstract async(): Crawler<Promise<O[]>>;

	abstract iterator(): Crawler<AsyncIterableIterator<O>>;

	abstract stream(): Crawler<Readable>;

	abstract sync(): Crawler<O[]>;
}

class CrawlerDownWithDirentBuilder extends CommonCrawlerBuilder<CrawlerDownOptions, Dirent> {
	constructor(options: CrawlerDownOptions) {
		super({
			...options,
			direction: 'down',
			returnType: 'Dirent',
		});
	}

	async(): Crawler<Promise<Dirent[]>> {
		return new DirentAsyncCrawler();
	}

	iterator(): Crawler<AsyncIterableIterator<Dirent>> {
		return new DirentIteratorCrawler();
	}

	stream(): Crawler<Readable> {
		return new DirentStreamCrawler();
	}

	sync(): Crawler<Dirent[]> {
		return new DirentSyncCrawler();
	}
}

class CrawlerDownBuilder extends CommonCrawlerBuilder<CrawlerDownOptions, string> {
	constructor() {
		super({
			direction: 'down',
			returnType: 'string',
		});
	}

	async(): Crawler<Promise<string[]>> {
		return new StringAsyncCrawler();
	}

	iterator(): Crawler<AsyncIterableIterator<string>> {
		return new StringIteratorCrawler();
	}

	stream(): Crawler<Readable> {
		return new StringStreamCrawler();
	}

	sync(): Crawler<string[]> {
		return new StringSyncCrawler();
	}

	withDirent(): CrawlerDownWithDirentBuilder {
		return new CrawlerDownWithDirentBuilder(this.options);
	}
}

export default abstract class CrawlerBuilder {
	static down(): CrawlerDownBuilder {
		return new CrawlerDownBuilder();
	}
}
