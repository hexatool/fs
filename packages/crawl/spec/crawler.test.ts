import type { Dirent } from 'node:fs';
import path from 'node:path';
import process from 'node:process';

import { describe, expect, it } from 'vitest';

import crawler from '../src/builder';
import { apiTypes, directions } from '../src/types';

let crawlModuleFilePath = path.relative(process.cwd(), path.resolve(__dirname, '..'));
crawlModuleFilePath = crawlModuleFilePath === '' ? '.' : crawlModuleFilePath;

describe('@hexatool/fs-crawl', () => {
	const CRAWL_CURRENT_DIR = 'crawl current directory';
	describe.each(directions)(CRAWL_CURRENT_DIR, direction => {
		describe.each(apiTypes)(`${CRAWL_CURRENT_DIR} ${direction}`, type => {
			if (type === 'stream') {
				it(`${type} ${CRAWL_CURRENT_DIR} ${direction}`, () =>
					new Promise<void>(done => {
						const files: (Dirent | string)[] = [];
						const stream = crawler.down().stream().start();
						stream.on('data', d => files.push(d as string));
						stream.on('end', () => {
							expect(files).toBeDefined();
							expect(files.length).toBeGreaterThan(0);
							expect(files.every(t => t)).toBeTruthy();
							done();
						});
					}));
			} else if (type === 'iterator') {
				it(`${type} ${CRAWL_CURRENT_DIR} ${direction}`, async () => {
					const files: string[] = [];
					const iterator = crawler.down().iterator().start(crawlModuleFilePath);

					for await (const item of iterator) {
						files.push(item);
					}
					expect(files).toBeDefined();
					expect(files.length).toBeGreaterThan(0);
					expect(files.every(t => t)).toBeTruthy();
				});
			} else {
				it(`${type} ${CRAWL_CURRENT_DIR} ${direction}`, async () => {
					const files =
						type === 'async'
							? await crawler.down().async().start()
							: crawler.down().sync().start();
					expect(files).toBeDefined();
					expect(files.length).toBeGreaterThan(0);
					expect(files.every(t => t)).toBeTruthy();
				});
			}
		});
	});

	const CRAWL_CURRENT_WITH_DIRENT = 'crawl current directory with Dirent';
	describe.each(directions)(CRAWL_CURRENT_WITH_DIRENT, direction => {
		describe.each(apiTypes)(`${CRAWL_CURRENT_WITH_DIRENT} ${direction}`, type => {
			if (type === 'stream') {
				it(`${type} ${CRAWL_CURRENT_WITH_DIRENT} ${direction}`, () =>
					new Promise<void>(done => {
						const files: (Dirent | string)[] = [];
						const stream = crawler.down().withDirent().stream().start();
						stream.on('data', d => files.push(d as string));
						stream.on('end', () => {
							expect(files).toBeDefined();
							expect(files.length).toBeGreaterThan(0);
							expect(files.every(t => t)).toBeTruthy();
							done();
						});
					}));
			} else if (type === 'iterator') {
				it(`${type} ${CRAWL_CURRENT_WITH_DIRENT} ${direction}`, async () => {
					const files: Dirent[] = [];
					const iterator = crawler
						.down()
						.withDirent()
						.iterator()
						.start(crawlModuleFilePath);

					for await (const item of iterator) {
						files.push(item);
					}
					expect(files).toBeDefined();
					expect(files.length).toBeGreaterThan(0);
					expect(files.every(t => t)).toBeTruthy();
				});
			} else {
				it(`${type} ${CRAWL_CURRENT_WITH_DIRENT} ${direction}`, async () => {
					const files =
						type === 'async'
							? await crawler.down().withDirent().async().start()
							: crawler.down().withDirent().sync().start();
					expect(files).toBeDefined();
					expect(files.length).toBeGreaterThan(0);
				});
			}
		});
	});

	const CRAWL_CUSTOM_DIR = 'crawl custom directory';
	describe.each(directions)(CRAWL_CUSTOM_DIR, direction => {
		describe.each(apiTypes)(`${CRAWL_CUSTOM_DIR} ${direction}`, type => {
			if (type === 'stream') {
				it(`${type} ${CRAWL_CUSTOM_DIR} ${direction}`, () =>
					new Promise<void>(done => {
						const files: string[] = [];
						const stream = crawler.down().stream().start(crawlModuleFilePath);
						stream.on('data', d => files.push(d as string));
						stream.on('end', () => {
							expect(files).toMatchSnapshot();
							done();
						});
					}));
			} else if (type === 'iterator') {
				it(`${type} ${CRAWL_CUSTOM_DIR} ${direction}`, async () => {
					const files: (Dirent | string)[] = [];
					const iterator = crawler.down().iterator().start(crawlModuleFilePath);

					for await (const item of iterator) {
						files.push(item);
					}
					expect(files).toMatchSnapshot();
				});
			} else {
				it(`${type} ${CRAWL_CUSTOM_DIR} ${direction}`, async () => {
					const files =
						type === 'async'
							? await crawler.down().async().start(crawlModuleFilePath)
							: crawler.down().sync().start(crawlModuleFilePath);
					expect(files).toMatchSnapshot();
				});
			}
		});
	});

	const CRAWL_CUSTOM_DIR_WITH_DIRENT = 'crawl custom directory with Dirent';
	describe.each(directions)(CRAWL_CUSTOM_DIR_WITH_DIRENT, direction => {
		describe.each(apiTypes)(`${CRAWL_CUSTOM_DIR_WITH_DIRENT} ${direction}`, type => {
			if (type === 'stream') {
				it(`${type} ${CRAWL_CUSTOM_DIR_WITH_DIRENT} ${direction}`, () =>
					new Promise<void>(done => {
						const files: (Dirent | string)[] = [];
						const stream = crawler
							.down()
							.withDirent()
							.stream()
							.start(crawlModuleFilePath);
						stream.on('data', d => files.push(d as string));
						stream.on('end', () => {
							expect(files).toMatchSnapshot();
							done();
						});
					}));
			} else if (type === 'iterator') {
				it(`${type} ${CRAWL_CUSTOM_DIR_WITH_DIRENT} ${direction}`, () => async () => {
					const files: Dirent[] = [];
					const iterator = crawler
						.down()
						.withDirent()
						.iterator()
						.start(crawlModuleFilePath);

					for await (const item of iterator) {
						files.push(item);
					}
					expect(files).toMatchSnapshot();
				});
			} else {
				it(`${type} ${CRAWL_CURRENT_WITH_DIRENT} ${direction}`, async () => {
					const files =
						type === 'async'
							? await crawler.down().withDirent().async().start(crawlModuleFilePath)
							: crawler.down().withDirent().sync().start(crawlModuleFilePath);
					expect(files).toMatchSnapshot();
				});
			}
		});
	});
});
