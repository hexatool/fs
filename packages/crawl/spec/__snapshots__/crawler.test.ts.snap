// Vitest Snapshot v1

exports[`@hexatool/fs-crawl > crawl custom directory > crawl custom directory down > async crawl custom directory down 1`] = `
[
  ".eslintignore",
  ".eslintrc.json",
  ".npmignore",
  "CHANGELOG.md",
  "LICENSE",
  "README.md",
  "package.json",
  "spec",
  "src",
  "tsconfig.json",
  "vite.config.ts",
]
`;

exports[`@hexatool/fs-crawl > crawl custom directory > crawl custom directory down > iterator crawl custom directory down 1`] = `
[
  ".eslintignore",
  ".eslintrc.json",
  ".npmignore",
  "CHANGELOG.md",
  "LICENSE",
  "README.md",
  "package.json",
  "spec",
  "src",
  "tsconfig.json",
  "vite.config.ts",
]
`;

exports[`@hexatool/fs-crawl > crawl custom directory > crawl custom directory down > stream crawl custom directory down 1`] = `
[
  ".eslintignore",
  ".eslintrc.json",
  ".npmignore",
  "CHANGELOG.md",
  "LICENSE",
  "README.md",
  "package.json",
  "spec",
  "src",
  "tsconfig.json",
  "vite.config.ts",
]
`;

exports[`@hexatool/fs-crawl > crawl custom directory > crawl custom directory down > sync crawl custom directory down 1`] = `
[
  ".eslintignore",
  ".eslintrc.json",
  ".npmignore",
  "CHANGELOG.md",
  "LICENSE",
  "README.md",
  "package.json",
  "spec",
  "src",
  "tsconfig.json",
  "vite.config.ts",
]
`;

exports[`@hexatool/fs-crawl > crawl custom directory with Dirent > crawl custom directory with Dirent down > async crawl current directory with Dirent down 1`] = `
[
  Dirent {
    "name": ".eslintignore",
    Symbol(type): 1,
  },
  Dirent {
    "name": ".eslintrc.json",
    Symbol(type): 1,
  },
  Dirent {
    "name": ".npmignore",
    Symbol(type): 1,
  },
  Dirent {
    "name": "CHANGELOG.md",
    Symbol(type): 1,
  },
  Dirent {
    "name": "LICENSE",
    Symbol(type): 1,
  },
  Dirent {
    "name": "README.md",
    Symbol(type): 1,
  },
  Dirent {
    "name": "package.json",
    Symbol(type): 1,
  },
  Dirent {
    "name": "spec",
    Symbol(type): 2,
  },
  Dirent {
    "name": "src",
    Symbol(type): 2,
  },
  Dirent {
    "name": "tsconfig.json",
    Symbol(type): 1,
  },
  Dirent {
    "name": "vite.config.ts",
    Symbol(type): 1,
  },
]
`;

exports[`@hexatool/fs-crawl > crawl custom directory with Dirent > crawl custom directory with Dirent down > stream crawl custom directory with Dirent down 1`] = `
[
  Dirent {
    "name": ".eslintignore",
    Symbol(type): 1,
  },
  Dirent {
    "name": ".eslintrc.json",
    Symbol(type): 1,
  },
  Dirent {
    "name": ".npmignore",
    Symbol(type): 1,
  },
  Dirent {
    "name": "CHANGELOG.md",
    Symbol(type): 1,
  },
  Dirent {
    "name": "LICENSE",
    Symbol(type): 1,
  },
  Dirent {
    "name": "README.md",
    Symbol(type): 1,
  },
  Dirent {
    "name": "package.json",
    Symbol(type): 1,
  },
  Dirent {
    "name": "spec",
    Symbol(type): 2,
  },
  Dirent {
    "name": "src",
    Symbol(type): 2,
  },
  Dirent {
    "name": "tsconfig.json",
    Symbol(type): 1,
  },
  Dirent {
    "name": "vite.config.ts",
    Symbol(type): 1,
  },
]
`;

exports[`@hexatool/fs-crawl > crawl custom directory with Dirent > crawl custom directory with Dirent down > sync crawl current directory with Dirent down 1`] = `
[
  Dirent {
    "name": ".eslintignore",
    Symbol(type): 1,
  },
  Dirent {
    "name": ".eslintrc.json",
    Symbol(type): 1,
  },
  Dirent {
    "name": ".npmignore",
    Symbol(type): 1,
  },
  Dirent {
    "name": "CHANGELOG.md",
    Symbol(type): 1,
  },
  Dirent {
    "name": "LICENSE",
    Symbol(type): 1,
  },
  Dirent {
    "name": "README.md",
    Symbol(type): 1,
  },
  Dirent {
    "name": "package.json",
    Symbol(type): 1,
  },
  Dirent {
    "name": "spec",
    Symbol(type): 2,
  },
  Dirent {
    "name": "src",
    Symbol(type): 2,
  },
  Dirent {
    "name": "tsconfig.json",
    Symbol(type): 1,
  },
  Dirent {
    "name": "vite.config.ts",
    Symbol(type): 1,
  },
]
`;
